src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -31,109 +30,134 @@
+            switch (entry.operation.toLowerCase()) {
+                case "add":
+                    clinicService.savePet(pet);
+                case "delete":
+                    // originally provided logic can be wrapped in this function
+                    clinicService.deletePet(pet);
             }
+            pets.add(pet);
+        }
+    }
 
-            if (csv.charAt(i) == ';') {
-                i++;
+    private HttpHeaders constructHttpHeaders(String headerValue){
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("errors", headerValue);
+        return headers;
+    }
 
-                field = "";
-                while (i < csv.length() && csv.charAt(i) != '\n') {
-                    field += csv.charAt(i++);
-                }
+    private Serializable toPetName(String name) {
+        if (name != "null") {
+            return name;
+        } else {
+            return constructHttpHeaders("Pet name is empty");
+        }
+    }
 
-                if (field.toLowerCase().equals("add")) {
-                    clinicService.savePet(pet);
-                } else {
-                    for (Pet q : pet.getOwner().getPets()) {
-                        if (q.getName().equals(pet.getName())) {
-                            if (q.getType().getId().equals(pet.getType().getId())) {
-                                if (pet.getBirthDate().equals(q.getBirthDate())) {
-                                    clinicService.deletePet(q);
-                                }
-                            }
-                        }
-                    }
-                }
-
-            } else {
-                clinicService.savePet(pet);
-            }
-            i++;
+    private Serializable toDate(String date) {
+        try {
+            return (new SimpleDateFormat("yyyy-MM-dd")).parse(date);
+        } catch (ParseException e) {
+            return constructHttpHeaders("date " + date + " not valid");
+        }
+    }
 
-            pets.add(pet);
+    private Object toPetType(String petTypeCSV) {
+        ArrayList<PetType> petTypes = (ArrayList<PetType>) clinicService.findPetTypes();
+        for (PetType petType : petTypes) {
+            if (petType.getName().equalsIgnoreCase(petTypeCSV)) {
+                return petType;
+            }
+        }
+        return constructHttpHeaders("Pet type does not exist");
+    }
 
-        } while (i < csv.length() && pet != null);
+    private Object toOwner(String name) {
+        List<Owner> matchingOwners = clinicService.findAllOwners()
+            .stream()
+            .filter(o -> o.getLastName().equals(owner))
+            .collect(Collectors.toList());
+
+        if (matchingOwners.size() == 0) {
+            return constructHttpHeaders("Owner not found");
+        }
+        // why should we throw error for owner not unique? Can't an owner have more than one pets at our clinic?
+        // Or two person with same family names?
```

Comment:
Two persons with the same family name. Usually this is solved by special identifiers instead of using the last name.